data_steward/analytics/cdr_ops/incremental_load_qc.py

```diff
@@ -0,0 +1,348 @@
+
+df = execute(client, union_all_query)
+
+success_null_check = (
+    "All OBSERVATION, SURVEY_CONDUCT, and PERSON records have "
+    "valid corresponding records in their mapping/ext tables. <br>"
+    "And there are no invalid records in the mapping/ext tables.")
+failure_null_check = (
+    "These mapping/ext tables are inconsistent with their "
+    "correspondants. <br>Look at these tables and investigate why "
+    "they are inconsistent: <b>{tables}</b><br><br>")
+
+render_message(df,
+               success_null_check,
+               failure_null_check,
+               failure_msg_args={'tables': ', '.join(df.table_name)})
+
+# -
+
+# ## QC 4. (Only for `is_deidentified == True`) Confirm `person_ext`'s state related columns come from `source_dataset`
+# `state_of_residence_concept_id` and `state_of_residence_source_value` in `incremental_dataset.person_ext` are
+# all NULL. This is because these two columns do not originate from the basics. <br>To have a complete `new_dataset.person_ext`,
+# we must pull these columns from `source_dataset`, not from `incremental_dataset`.
+
+# +
+if is_deidentified == "False":
+    success_msg = "Skipping this check person_ext table exists only in DEID datasets.<br>"
+    render_message('', success_msg=success_msg)
+
+else:
+    query = JINJA_ENV.from_string('''
+        SELECT *
+        FROM `{{project}}.{{new_dataset}}.person_ext` n
+        JOIN `{{project}}.{{source_dataset}}.person_ext` s
+        ON n.person_id = s.person_id
+        WHERE (
+            n.state_of_residence_concept_id != s.state_of_residence_concept_id
+            OR n.state_of_residence_source_value != s.state_of_residence_source_value
+        )
+        OR (
+            (n.state_of_residence_concept_id IS NULL AND s.state_of_residence_concept_id IS NOT NULL)
+            OR (n.state_of_residence_source_value IS NULL AND s.state_of_residence_source_value IS NOT NULL)
+        )
+    ''').render(project=project_id,
+                new_dataset=new_dataset,
+                source_dataset=source_dataset)
+
+    df = execute(client, query)
+
+    success_null_check = (
+        f"All state columns in {new_dataset}.person_ext come from {source_dataset}, "
+        f"not {incremental_dataset}.<br><br>")
+    failure_null_check = (
+        f"There are <b>{len(df)}</b> records in {new_dataset}.person_ext that have "
+        f"unmatching state columns from {source_dataset}. <br>Look at the table and "
+        "investigate why they are inconsistent.<br><br>")
+
+    render_message(df, success_null_check, failure_null_check)
+
+# -
+
+# ## QC 5. Confirm `SURVEY_CONDUCT` references `OBSERVATION` correctly
+# This hotfix runs delete and insert on both `SURVEY_CONDUCT` and `OBSERVATION`. <br>
+# `survey_conduct_id` must have corresponding `questionnaire_response_id` in `OBSERVATION`.
+# This QC confirms all `SURVEY_CONDUCT` records have corresponding records in `OBSERVATION`
+# after the hotfix.
+
+# +
```

Comment:
This does check if `survey_conduct_id` is in the set of `questionnaire_response_id`s.  I think we should also check whether `questionniare_response_id` is in the set of `survey_conduct_id`s.

Additionally, can we add a check to tell us how many new `questionnaire_response_id`s and `survey_conduct_id`s come from the "increment"?

Reply:
I added the additional checks with this commit: https://github.com/all-of-us/curation/pull/1489/commits/d6dcfa387e32c1755fe0fb840bfd826bd64ef8c2